### PR TITLE
Bugfix: Fix marking TV Shows as watched/unwatched

### PR DESCRIPTION
**Pull request by wutschel**
## Description

As reported in [this forum post](https://forum.kodi.tv/showthread.php?tid=359717&pid=3106513#pid3106513) it is not possible to mark a complete TV Show as watched/unwatched with Kodi 19.4. Even though the App is sending the valid command

`{"jsonrpc":"2.0","method":"VideoLibrary.SetTVShowDetails","params":{"tvshowid":11, "playcount":1}, "id":2}`

and receives the response `"OK"` from the API, Kodi server does not change the playcount.

This PR implements a solution similar to Chorus Web Interface which simply sets the playcount for each episode of a TV Show. As I am not sure if Kodi server itself should be fixed (in case the JSON command from above should suffice), or if the App shall work around this, I am keeping this draft for now.

## Summary for release notes

Bugfix: Fix marking TV Shows as watched/unwatched

### CODE REVIEW

**Comment by wutschel** [comment]
@kambala-decapitator, you can already review the code changes. As there is no feedback on other ways to implement this I tend to bring in this change.

**Comment by wutschel** [comment]
To document some additional information here:
- Chorus updates `playcount` per episode when marking a TV Show as _watched_.
- Kore (Android remote) does not seem to support this feature.

File: XBMC Remote/DetailViewController.m [hunk 1]
```diff
@@ -3274,7 +3274,55 @@ - (void)markVideo:(NSMutableDictionary*)item indexPath:(NSIndexPath*)indexPath w
     NSString *methodToCall = @"";
     NSString *family = item[@"family"];
     if ([family isEqualToString:@"tvshowid"]) {
-        methodToCall = @"VideoLibrary.SetTVShowDetails";
+        methodToCall = @"VideoLibrary.SetEpisodeDetails";
+        NSDictionary *params = [NSDictionary dictionaryWithObjectsAndKeys:item[@"tvshowid"], @"tvshowid", @[@"season", @"episode"], @"properties", nil];
```
Comment (kambala-decapitator) [inline XBMC Remote/DetailViewController.m:3273]: I guess `item[@"tvshowid"]` can’t be nil at this point:
```suggestion
        NSDictionary *params = @{
            @"tvshowid": item[@"tvshowid"],
            @"properties": @[@"season", @"episode"],
        };
```

File: XBMC Remote/DetailViewController.m [hunk 2]
```diff
@@ -3274,7 +3274,55 @@ - (void)markVideo:(NSMutableDictionary*)item indexPath:(NSIndexPath*)indexPath w
...
+         callMethod:@"VideoLibrary.GetEpisodes"
+         withParameters:params
+         onCompletion:^(NSString *methodName, NSInteger callId, id methodResult, DSJSONRPCError *methodError, NSError *error) {
+            if (error == nil && methodError == nil) {
+                // Set the playcount for each episode of the TV Show (fire-and-forget, no error check)
+                for (id arrayItem in methodResult[@"episodes"]) {
+                    NSDictionary *params = [NSDictionary dictionaryWithObjectsAndKeys:
+                                            arrayItem[@"episodeid"], @"episodeid",
+                                            @(watched), @"playcount",
+                                            nil];
```
Comment (kambala-decapitator) [inline XBMC Remote/DetailViewController.m:3284]: I guess `arrayItem[@"episodeid"]` can’t be nil at this point:
```suggestion
                    NSDictionary *params = @{
                        @"episodeid": arrayItem[@"episodeid"],
                        @"playcount": @(watched),
                    };
```

File: XBMC Remote/DetailViewController.m [hunk 3]
```diff
@@ -3274,7 +3274,55 @@ - (void)markVideo:(NSMutableDictionary*)item indexPath:(NSIndexPath*)indexPath w
...
+                    [[Utilities getJsonRPC]
+                     callMethod:methodToCall
+                     withParameters:params
+                     onCompletion:^(NSString *methodName, NSInteger callId, id methodResult, DSJSONRPCError *methodError, NSError* error) {}];
```
Comment (kambala-decapitator) [inline XBMC Remote/DetailViewController.m:3288]: can you pass nil instead of empty block?

File: XBMC Remote/DetailViewController.m [hunk 4]
```diff
@@ -3274,7 +3274,55 @@ - (void)markVideo:(NSMutableDictionary*)item indexPath:(NSIndexPath*)indexPath w
...
+                mainMenu *menuItem = [self getMainMenu:item];
+                NSDictionary *parameters = [Utilities indexKeyedDictionaryFromArray:menuItem.mainParameters[choosedTab]];
+                NSMutableDictionary *mutableParameters = [parameters[@"parameters"] mutableCopy];
+                NSMutableArray *mutableProperties = [parameters[@"parameters"][@"properties"] mutableCopy];
+                if ([parameters[@"FrodoExtraArt"] boolValue] && AppDelegate.instance.serverVersion > 11) {
+                    [mutableProperties addObject:@"art"];
+                    mutableParameters[@"properties"] = mutableProperties;
+                }
+                if (mutableParameters[@"file_properties"] != nil) {
+                    mutableParameters[@"properties"] = mutableParameters[@"file_properties"];
```
Comment (kambala-decapitator) [inline XBMC Remote/DetailViewController.m]: is it correct to overwrite what was set a couple of lines above?

**Review by kambala-decapitator — APPROVED** [review]

File: XBMC Remote/DetailViewController.m (hunk 2 above)
Comment (wutschel) [reply XBMC Remote/DetailViewController.m:3284]: I cannot guarantee the API response is complete. So, would like to keep it. The condition checks another key.

File: XBMC Remote/DetailViewController.m (hunk 3 above)
Comment (wutschel) [reply XBMC Remote/DetailViewController.m:3288]: Agreed

File: XBMC Remote/DetailViewController.m (hunk 4 above)
Comment (wutschel) [reply XBMC Remote/DetailViewController.m]: Oh, I see. Wasn't even looking at this as I just moved duplicate code. Let us please keep as is for now. Do nt want to risk any regression.

**Review by kambala-decapitator — APPROVED** [review]

**Comment by wutschel** [comment]
Squashed and rebased to master.

**Comment by kambala-decapitator** [comment]
don't you want to squash https://github.com/xbmc/Official-Kodi-Remote-iOS/pull/700/commits/8008f655abbc1329048284698e0c25d5aee5c3d2 too? :)

**Comment by wutschel** [comment]
> don't you want to squash [8008f65](https://github.com/xbmc/Official-Kodi-Remote-iOS/commit/8008f655abbc1329048284698e0c25d5aee5c3d2) too? :)

Thanks, was forgetting this as the fixup was added a while back.

**Pull request merged**